Implement Pages 3.1 and Create Tests

volosied opened the pull request:
fixes #21743 - failed previously due to the EL Resolver changes, but now they work.
fixes #20735 - main change added 
partly addresses #20737 via the new EL Resolvers list.  Follow-up issue: # https://github.com/OpenLiberty/open-liberty/issues/22346
fixes #20738 - logs a warning when this directive is used & synchronizes service method as suggested in the spec issue
fixes #22325 - jsp-plugin is now a no-op 
fixes #20432 -  JspPropertyGroupDescriptor#getErrorOnELNotFound() is handled 

This code implements the following items:
- Correct EL Resolver List for Page 3.1 ( ImportResolver, ELNotFoundResolver) 
- errorOnELNotFound option as the following three types: 
  - Page Directive
  - Tag Directive
  - JSP Property Group (required changes to the com.ibm.ws.javaee.ddmodel & com.ibm.ws.javaee.dd bundles) 
  - Pages 3.1 ignores jsp:plugin 
  - isThreadSafe synchronizes the service method in Page 3.1 and logs a warning when it's used (isThreadSafe is deprecated and planned for removal in 4.0) 
  
The Pages 3.1 feature requires a new bundle: io.openliberty.pages.3.1.internal.factories which is mostly a copy of com.ibm.ws.jsp.2.3 (but with specific changes to the EL Resolvers & the errorOnELNotFound option) 

We had some messaging trouble when using `CWWJS`, so note the following issue: #22421. We will need the header to be uploaded for the other translation files.

volosied committed:
Implement Pages 3.1 and Create Tests

Corrections

Fix up io.openliberty.pages.3.1.internal.factories bundle name:

Revert package names in 3.1 factory bundle

Fix if nesting

Add tag directive apps

Implement option in jsp tags

test updates

Reomve method from JspClassInformation

Add CWWJS0002 message -- jsp.error.page.invalid.erroronelnotfound

Remove uneccessary check

Add Pages3.1+ check

Update versions for feature checker

Remove commented code

volosied commented:
#build

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/validator/ValidateVisitor.java:
```diff
@@ -1199,6 +1200,9 @@ protected void validateAttributes(Element jspElement, String[] validAttrs) throw
                     break;
                 }
             }
+            if(attr.getNodeName().equals("errorOnELNotFound") && PagesVersionHandler.isPages31OrHigherLoaded()){
+                valid = true;     
+            }
```
remove; just add errorOnELNotFound to the list

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/validator/ValidateTagFileVisitor.java:
```diff
@@ -195,6 +196,22 @@ else if (!jspConfiguration.getDeferredSyntaxAllowedAsLiteral().equals(directiveV
+                          jspConfiguration.setErrorOnELNotFound(false);
+                      }
+                      else
+                          throw new JspTranslationException(jspElement, "jsp.error.page.invalid.iselignored");
```
add message for jsp.error.page.invalid.iselignored

volosied commented:
#build

volosied commented on dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/runtime/JspContextWrapper.java:
```diff
@@ -0,0 +1,514 @@
+    public ELContext getELContext() {
+        ELContext context =  this.invokingJspCtxt.getELContext();
+
+        if(tag instanceof com.ibm.ws.jsp.runtime.JspDirectiveInfo && ((com.ibm.ws.jsp.runtime.JspDirectiveInfo)tag).isErrorOnELNotFound() ){
+            context.putContext(jakarta.servlet.jsp.el.NotFoundELResolver.class, true);
+        }
```
3.1 specific code

volosied commented on dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/el/JasperELResolver.java:
```diff
@@ -0,0 +1,174 @@
+        add(new ListELResolver());
+        add(new ArrayELResolver());
+        add(new BeanELResolver());
+        add(new ScopedAttributeELResolver());
+        add(new ImportELResolver());
+        add(new NotFoundELResolver());;
```
Pages 3.1 specific code

tbitonti reviewed (commented):
Looks good to me, excepting:

dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/messages.nlsprops
- "Invalid" is not to be used, per messaging guidelines.
- Missing explanation and user action.
```
jsp.error.page.invalid.erroronelnotfound=CWWJS0002: Invalid value for errorOnELNotFound attribute
jsp.error.page.invalid.erroronelnotfound.explanation=
jsp.error.page.invalid.erroronelnotfound.useraction=
```
There are a few files with missing copyright updates, and a few other minor comments.  See:
[review_22130_23-Aug-2022.txt](https://github.com/OpenLiberty/open-liberty/files/9404384/review_22130_23-Aug-2022.txt)

volosied committed:
Update to isThreadSafe for Pages 3.1

Ignore jsp:plugin for Pages 3.1

volosied committed:
Review Corrections

Copyright updates to io.openliberty.pages.3.1.internal.factories

Few other corrections

Change back to CWWJS0003 to avoid [ant:validate]   (fatal) jsp.isthreadsafe.warning uses an unregistered component prefix (CWWJS)

More Corrections & Updates

Fix bnd & remove comment

volosied commented:
#build

volosied commented:
#build

pnicolucci commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/configuration/JspConfigPropertyGroup.java:
```diff
@@ -18,103 +18,16 @@
+     * This method is under EE10's JspPropertyGroupDescriptor, but we are using the older servlet APIs to build 
+     * this bundle. This method has been added under JspConfigPropertyGroup instead. 
+     * It will only be used when Pages 3.1 is enabled. 
+     * Will the use of an unknown identifier in EL within a JSP page trigger an error for this group?
```
Can we avoid the use of "EL / JSP" here in the comments? If specific to EE10  we should be good to use "Jakarta Server Pages page, Expression Language"

volosied replied:
I copied this from the [spec](https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/descriptor/jsppropertygroupdescriptor#getErrorOnELNotFound()), but, yes, I can update it.

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateJspVisitor.java:
```diff
@@ -191,73 +195,95 @@ else if (methodWriter instanceof CleanupTaglibLookupWriter) {
+         *  For Page 3.1, the jsp service method is synchronized instead (performance hit)
+         *  isThreadSafe is not recommended anymore. 
+         */ 
+        if(PagesVersionHandler.isPages30OrLowerLoaded()){
+            boolean singleThreaded = validatorResult.isSingleThreaded();
+
+            if (singleThreaded)
+                interfaces.add("SingleThreadModel");
+        } else {
+            logger.logp(Level.WARNING, CLASS_NAME, "generateClassSection", "jsp.isthreadsafe.warning");
+        }
+
```
I think the logic here is bad. Needs to be re-done.

pnicolucci replied:
Since we are in common code here we should clean it up before merging IMO. What part of the logic specifically are you worried about?

volosied replied:
I think I need to add a singleThreaded check before logging the warning.  It looks like it would be logged every time the pages 3.1 feature is used.

pnicolucci commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/configuration/JspConfiguration.java:
```diff
@@ -316,6 +352,8 @@ public String toString(){
 		 	"elIgnored =                          [" + elIgnored +"]"+separatorString+
 		 	"elIgnoredSetTrueInPropGrp =          [" + elIgnoredSetTrueInPropGrp +"]"+separatorString+
 		 	"elIgnoredSetTrueInPage =             [" + elIgnoredSetTrueInPage +"]"+separatorString+
+            "errorOnELNotFoundInPropGroup =       [" + errorOnELNotFoundInPropGroup +"]"+separatorString+
```
Can you clean up the formatting here?

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateJspVisitor.java:
```diff
@@ -275,59 +301,71 @@ protected void generateStaticSection() {
+         *  https://github.com/jakartaee/pages/issues/206#issuecomment-934272204
+         */
+        String sync = "";
+        if(PagesVersionHandler.isPages31OrHigherLoaded()){
+            if(validatorResult.isSingleThreaded()){
+                sync = "synchronized";
+            }
+        }
+
         writer.println(
-            "public void "
+            "public " + sync + " void "
```
Service method is synchronized when  isThreadSafe = false

pnicolucci commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/configuration/JspConfigurationManager.java:
```diff
@@ -156,6 +157,10 @@ public synchronized JspConfiguration getConfigurationForUrl(String url) {
 
+                if(PagesVersionHandler.isPages31OrHigherLoaded()){
+                    if(propertyGroup.getErrorOnELNotFound()!=null)
+                    configuration.setErrorOnELNotFound(Boolean.valueOf(propertyGroup.getErrorOnELNotFound()));
```
Add some indentation here

pnicolucci commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateJspVisitor.java:
```diff
@@ -275,59 +301,71 @@ protected void generateStaticSection() {
         writer.println();
+
+        /*
+         *  Mark recommened using synchonized, but with a performance caveat
```
What do you mean "Mark recommended", are we going to document this someplace? Do we have an issue opened if that is the case?

volosied replied:
It's all part of https://github.com/OpenLiberty/open-liberty/issues/20738.

isThreadSafe is deprecated, but deprecated means the behavior is planned for removal.  However, servlet 6.0 removed the SingleThreadModel.   So this behavior is removed after all.    So, by adding the synchronized keyword, it would try to keep the isThreadSafe behavior as before (but just without the SingleThreadModel interface).

volosied replied:
As Mark said, we're trying to achieve the same result without SingleThreadModel. See `https://github.com/jakartaee/pages/issues/206#issuecomment-934272204`  and the the comment above.

pnicolucci commented on dev/io.openliberty.pages.3.1.internal.factories/bnd.bnd:
```diff
@@ -0,0 +1,105 @@
+#*******************************************************************************
+# Copyright (c) 2017, 2022 IBM Corporation and others.
```
A 2017 copyright doesn't seem correct here for a pages 3.1 bundle new to EE10

pnicolucci commented on dev/io.openliberty.pages.3.1.internal.factories/bnd.bnd:
```diff
@@ -0,0 +1,105 @@
+
+Export-Package: \
+	com.ibm.ws.jsp.speclevel, \
+	com.ibm.ws.jsp23.webcontainerext;version="2.3", \
```
A 2.3 version seems odd, lets discuss. Also the naming of the pages if in the 3.1 bundle is also odd to include jsp2.3.

volosied replied:
Oops. I will fix this.  I think I'll keep the com.ibm.ws. prefix  ( as I had issues when I tried to switch it to io.openliberty ), but changes jsp23 -> jsp31 and the version to 3.1

epj commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/JspVisitor.java:
```diff
@@ -324,9 +325,15 @@ else if (jspElementType.equals(Constants.JSP_SETPROPERTY_TYPE)) {
+                        visitJspPluginEnd(jspElement);
+                    } else {
+                        if(com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled()&&logger.isLoggable(Level.FINEST)){
+                            logger.logp(Level.FINEST, CLASS_NAME, "processJspElement","Skipping jsp:plugin.");
```
Can you expand the message to say why the skip is happening?

volosied replied:
`jsp:plugin `is now a no-op as current browsers don't support the generated html tags `https://github.com/jakartaee/pages/issues/226`

epj replied:
I wasn't questioning the logic, just thought a few words could be added to the trace message to say its a no-op due to version.

volosied replied:
Oh, sorry I completely misread your comment. Yes, I can expand the message.

volosied replied:
Updated to: `Skipping jsp:plugin element as it is a no-op for Pages 3.1`

epj reviewed (changes requested):
just one comment for now

volosied committed:
Bump org.apache.jasper versions for 3.1 factory bundle

Update JspVistor

Fix copyright

volosied commented:
#build

volosied commented on dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/JspVisitor.java:
```diff
@@ -324,9 +325,15 @@ else if (jspElementType.equals(Constants.JSP_SETPROPERTY_TYPE)) {
+                        processChildren(jspElement);
+                        visitJspPluginEnd(jspElement);
+                    } else {
+                        if(com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled()&&logger.isLoggable(Level.FINEST)){
+                            logger.logp(Level.FINEST, CLASS_NAME, "processJspElement","Skipping jsp:plugin element as it is a no-op for Pages 3.1+");
+                        }
+                    }
```
Update this to a warning

volosied replied:
Kept as logging, not a warning. If used as  a warning, the logs filled out up numerous entries of this message since the method is entered many times

pnicolucci commented:
@volosied  comments  from our code review: 

1) Create tests for each function that you have implemented.
2) For jsp:plugin use a warning similar to isThreadSafe to be consistent
3) Check the UFO and ensure that we call out the jsp:plugin and isThreadSafe for the migration toolkit / deprecation if it makes sense.
4) Link this PR to https://github.com/OpenLiberty/open-liberty/issues/22325 and https://github.com/openliberty/open-liberty/issues/20738
5) can you just double check: https://github.com/openliberty/open-liberty/issues/20734 to make sure nothing in the impl is necessary to change?

volosied committed:
Add tests for jsp:plugin & isThreadSafe

volosied committed:
Mark getFeatureDescriptors as Deprecated

Update messages to warnings / errors

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented:
@pnicolucci 

1) Tests added to check for appropriate messages in the logs for isThreadSafe & jsp:plugin 
2) "jsp:plugin" logging is unchanged (logged only in trace) as method containing the logging is called numerous times. 
3) I updated the UFO to mention jsp:plugin & isThreadSafe in the migration issues slide.
4) This PR is linked to both issues. 
5) The getFeatureDescriptors() is used in the 3.1 factories bundle and I've mark it as deprecated. We can handle it's removal in the next Pages release.

volosied commented:
#libby

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

volosied commented on dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/messages.nlsprops:
```diff
@@ -1259,14 +1259,14 @@ system.java.compiler.not.found=JSPG0304W: The JSP attribute, useJDKCompiler, is
 system.java.compiler.not.found.useraction=Ensure that the Java SDK jar file, tools.jar, is in the class path and verify that the Java SDK supports the Java Compiler API. See Java Specification Request 199 for more information.
 
-jsp.feature.not.loaded.correctly=CWWJS0001: The JSP [Pages] feature did not load with an acceptable version.
+jsp.feature.not.loaded.correctly=CWWJS0001E: The JSP [Pages] feature did not load with an acceptable version.
```
This is the older message, but it never had the "E". I added it here now.

helyarp commented on dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/messages.nlsprops:
```diff
@@ -1259,6 +1259,14 @@ system.java.compiler.not.found=JSPG0304W: The JSP attribute, useJDKCompiler, is
+jsp.feature.not.loaded.correctly=CWWJS0001E: The JSP [Pages] feature did not load with an acceptable version.
 jsp.feature.not.loaded.correctly.explanation=The JSP [Pages] feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that JSP [Pages] feature. The feature was loaded with no version defined.
-jsp.feature.not.loaded.correctly.useraction=Check the server.xml file for existence of either a JSP [Pages] feature. Only one should be specified.
+jsp.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple JSP [Pages] features. Only one should be specified.
```
A nit: To remove the "should," you can change the second sentence to, for example, "The file must specify no more than one Pages feature."

helyarp commented on dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/messages.nlsprops:
```diff
@@ -1259,6 +1259,14 @@ system.java.compiler.not.found=JSPG0304W: The JSP attribute, useJDKCompiler, is
+jsp.error.page.invalid.erroronelnotfound.explanation=
+jsp.error.page.invalid.erroronelnotfound.useraction=
+
+jsp.isthreadsafe.warning=CWWJS0003W: Per the Pages 3.1 specification, pages developers are strongly advised not to use the deprecated page directive isThreadSafe. 
```
Suggested revision to remove the "per" and "strongly":

"The Pages 3.1 specification advises Pages developers not to use the deprecated page directive isThreadSafe."

Please change "Pages developer" back to "pages developer" if the latter has the preferred capitalization.

helyarp commented on dev/com.ibm.ws.jsp/resources/com/ibm/ws/jsp/resources/messages.nlsprops:
```diff
@@ -1259,6 +1259,14 @@ system.java.compiler.not.found=JSPG0304W: The JSP attribute, useJDKCompiler, is
+jsp.error.page.invalid.erroronelnotfound.useraction=
+
+jsp.isthreadsafe.warning=CWWJS0003W: Per the Pages 3.1 specification, pages developers are strongly advised not to use the deprecated page directive isThreadSafe. 
+jsp.isthreadsafe.warning.explanation=SingleThreadModel has been removed from Servlet 6.0. Pages 3.1 has deprecated isThreadSafe with plans to remove it in Page 4.0.
```
Suggested change to remove "has" and add "s" to "Page":

"SingleThreadModel was removed from Servlet 6.0. Pages 3.1 deprecated isThreadSafe with plans to remove it in Pages 4.0."

pnicolucci replied:
@volosied is this a certainty for Pages 4.0? Should we be more generic and say "in future Pages update"? Or perhaps remove the second part of the sentence? If it is deprecated, it is implied that it will be removed in a future release. I just don't want us to have to continue to come back and update this message if something doesn't make a specific release.

helyarp reviewed (commented):
Suggested changes

posmith commented:
L2 message review completed

volosied committed:
Message Review Updates

volosied commented:
@helyarp  I've updated the messages, Please take another look

helyarp reviewed (commented):
Changes look good!

volosied commented:
#build

epj reviewed (approved):
good work

volosied commented:
#libby
